### ISSUE TITLE
Change int boolean flags into bool (aka _Bool) for better clarity

### ISSUE THREAD

**Issue by seishinryohosha** [neomutt/neomutt#774]
* Why do you not like the current state?
  Using int for boolean flags is not very intuitive in a method prototype.
* What would you like to see to change?
  Change all int flags into bool flags.
  
I would start with changing all headers alphabetically and make for each letter 
a unique branch for easier reviewing.
After changing all top level headers I would go alphabetically in each folder.

For example the first branch would be `devel/switch_to_bool_a`:

So the branch would change:
* account.h
* address.h
* alias.h
* attach.h
* all calls to the prototypes in other files

Now I have some questions before I would be able to start:

1. Is the breakdown appropriate? 
    Or is it too granular or too shallow?
2. Are some files using int intentionally? 
    E.g. planning on using int in the future for other reasons, except boolean flags?
3. Other points I might be missing right now?

**Comment by flatcap** [comment]
Yes, it's a good idea.  But a little backstory first...

Even before we started NeoMutt, Mutt had a C99 requirement in the build.
This should allow:
- `bool`
- local for variables: `for (int i = 0; ...`
- variables declared nearer their use

Mutt didn't make use of any of these.
Using a lot of unix- and vim-fu, I made a lot of changes, I:
- changed the type of variables in functions to `bool`
- shrank the scope of `for` variables

and fixed the naming of:
- Global variables
- `struct`s
- function names

Besides vim, I also used [coccinelle](https://github.com/neomutt/coccinelle) -- a code transformation tool.

However careful I was, each time I managed to **** something up.
After days of staring at changes, you become blind to mistakes.

**Tip:** Keep the changes small.  I'd suggest one commit per prototype changed.
It will be a pain to organise, but it will mean we can divide up the job of reviewing.


If you're a CS student, you may be able to use Coccinelle, or Clang's AST to automate the changes.
If so, please teach us how :-)

**Comment by seishinryohosha** [comment]
I definitely look into the tools!
My proposed strategy would be as follow:
Follow the alphabetical PR versioning, but only for each header a single PR and doing a commit for each prototype a single commit.
This approach would even further divide it, and make the review easier.

First only change `int -> bool`. After that, start a new iteration with for example shrank the scope of `for` variables.

I looked into `account.h` and noticed that all functions, which return `int` only return `-1` or `0`.
Is it planned to change them also to `bool` or just leave them?
A quick `grep` on `mutt_account_fromurl` only used it in a boolean way.

Pro on changing them to `boo`:
* Makes it clear, that it should be used in a boolean context
* Only allows to values, `true == successfull` `false == failure`
* Makes it more consistent and easier for `if-clauses`

Contra:
* Enforces other verbose error notifications like `errno` or a `struct Buffer err` or something alike
  (This can be also a pro fact)
* Needs a lot of changes
* Is not extensible, e.g. someone wants to return different Codes, like `-1, -3 or even positives`

What's your opinion?

**Comment by flatcap** [comment]
> I looked into account.h and noticed that all functions, which return int only return -1 or 0.
> Is it planned to change them also to bool or just leave them?

A while back, we converted all the functions whose names _sounded_ boolean, `can_X()`, or `is_Y()`
The rest, we decided to leave them as they are.

If you find any functions that you think should be boolean, let us know.
We might consider changing their names.

> errno or a struct Buffer err

The error handling in Mutt isn't consistent.
If you think of any improvements, we'd love to hear them (as a separate issue).

Good luck and thanks :-)

If you want to chat about things, we're often in #neomutt on irc.freenode.net
